addons/stock/models/stock_rule.py

```diff
@@ -167,15 +172,57 @@ def _push_prepare_move_copy_values(self, move_to_copy, new_date):
         }
         return new_move_vals
 
+    def _get_procure_method(self, product, qty_needed, location):
+        """ When the procure_method of the rule is 'mts_else_mto', we need to get the quantity forcasted to know if
+        we have to apply MTS or MTO on the moves of the procurement.
+        :param product: The product of the move
+        :param qty_needed: The amount needed of that product
+        :param location: The location source of the product
+        :return: The procure_method to apply on the move
+        """
+        if self.procure_method == 'mts_else_mto':
+            qty_available = product.with_context(location=location.id).virtual_available
+            if float_compare(qty_needed, qty_available, precision_rounding=product.uom_id.rounding) <= 0:
+                return 'make_to_stock'
+            return 'make_to_order'
+        return self.procure_method
+
     @api.model
     def _run_pull(self, procurements):
         moves_values_by_company = defaultdict(list)
+        products_by_locations = defaultdict(list)
         for procurement, rule in procurements:
             if not rule.location_src_id:
                 msg = _('No source location defined on stock rule: %s!') % (rule.name, )
                 raise UserError(msg)
 
-            moves_values_by_company[procurement.company_id.id].append(rule._get_stock_move_values(*procurement))
+            if rule.procure_method == 'mts_else_mto':
+                products_by_locations[rule.location_src_id].append(procurement.product_id.id)
+
+        forecasted_qties_by_loc = {}
+        for location, product_ids in products_by_locations.items():
+            forecasted_qties = self.env['product.product'].browse(product_ids).with_context(location=location.id).read(['virtual_available'])
+            forecasted_qties_dict = {}
+            for qty_dict in forecasted_qties:
+                forecasted_qties_dict[qty_dict['id']] = qty_dict['virtual_available']
+            forecasted_qties_by_loc[location] = forecasted_qties_dict
+
+        for procurement, rule in procurements:
+            procure_method = rule.procure_method
+            if rule.procure_method == 'mts_else_mto':
+                qty_needed = procurement.product_uom._compute_quantity(procurement.product_qty, procurement.product_id.uom_id)
+                qty_available = forecasted_qties_by_loc[rule.location_src_id][procurement.product_id.id]
+                if float_compare(qty_needed, qty_available, precision_rounding=procurement.product_id.uom_id.rounding) <= 0:
+                    procure_method = 'make_to_stock'
+                    forecasted_qties_by_loc[rule.location_src_id][procurement.product_id.id] -= qty_needed
+                else:
+                    procure_method = 'make_to_order'
+
+            altered_rule = rule.copy()
+            altered_rule.procure_method = procure_method
+            move_values = altered_rule._get_stock_move_values(*procurement)
```

Comment:
try just after calling this
move_values = dict(move_values, procure_method=procure_method)